### PR TITLE
DE38655: Intermittent failure in GraphResourceRepositoryTest#testUpdateAttachedEntity

### PR DESCRIPTION
**Pull request by FrankGasparovic**

### CODE REVIEW

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 1]
```diff
@@ -631,4 +650,24 @@ private ResourceEntity persistResourceToZoneAndAssert(final ZoneEntity zoneEntit
         assertThat(this.resourceRepository.findOne(resourceEntity.getId()), equalTo(resource));
         return resourceEntity;
     }
+
+    @SuppressWarnings("unchecked")
+    static void deleteTwoLevelEntityAndParents(ZonableEntity e, final ZoneEntity zone,
```
Comment (anubhavi25) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: All parameters should be `final`. Also introduce generics into the method signature to allow you to remove the `unchecked` warning:

```
static <E extends ZonableEntity> void deleteTwoLevelEntityAndParents(final E zonableEntity, final ZoneEntity zone, final GraphGenericRepository<E> repository) {
    ...
}
```

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java [hunk 2]
```diff
@@ -631,4 +650,24 @@ private ResourceEntity persistResourceToZoneAndAssert(final ZoneEntity zoneEntit
...
+            final GraphGenericRepository repository) {
+        Set<Parent> parents = e.getParents();
+        for (Parent parent : parents) {
+            repository.delete(repository.getEntity(zone, parent.getIdentifier()));
+        }
+        repository.delete(e);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static void deleteThreeLevelEntityAndParents(ZonableEntity e, ZoneEntity zone,
```
Comment (anubhavi25) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: All parameters should be `final`. Also introduce generics into the method signature to allow you to remove the `unchecked` warning:

```
private static <E extends ZonableEntity> void deleteThreeLevelEntityAndParents(final E zonableEntity, final ZoneEntity zone, final GraphGenericRepository<E> repository) {
    ...
}
```

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java [hunk 3]
```diff
@@ -143,15 +148,19 @@ public void testParentAndChildAttributeSameNameDifferentValues() {
...
             invocationCount = CONCURRENT_TEST_INVOCATIONS)
     public void testSave() {
-        String subjectId = persistRandomSubjectToZone1AndAssert().getSubjectIdentifier();
+        SubjectEntity subjectEntity = persistRandomSubjectToZone1AndAssert();
+        String subjectId = subjectEntity.getSubjectIdentifier();
         GraphTraversal<Vertex, Vertex> traversal = this.graphTraversalSource.V().has(SUBJECT_ID_KEY, subjectId);
         assertThat(traversal.hasNext(), equalTo(true));
         assertThat(traversal.next().property(SUBJECT_ID_KEY).value(), equalTo(subjectId));
+        this.subjectRepository.delete(subjectEntity);
+
```
Comment (anubhavi25) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Remove empty line

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java [hunk 4]
```diff
@@ -227,6 +239,8 @@ public void testGetSubjectEntityAndDescendantsIds() {
         descendantsIds = this.subjectRepository
                 .getSubjectEntityAndDescendantsIds(new SubjectEntity(TEST_ZONE_1, "/nonexistent-subject"));
         assertThat(descendantsIds, empty());
+        GraphResourceRepositoryTest.deleteTwoLevelEntityAndParents(agentMulder, TEST_ZONE_1, this.subjectRepository);
+
```
Comment (anubhavi25) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Remove empty line

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java [hunk 5]
```diff
@@ -183,6 +193,8 @@ public void testSaveScopes() {
         }
         assertThat(parent, notNullValue());
         assertThat("Expected scope not found on subject.", parent.getScopes().contains(SITE_BASEMENT));
+        GraphResourceRepositoryTest.deleteTwoLevelEntityAndParents(subject, TEST_ZONE_1, this.subjectRepository);
+
```
Comment (anubhavi25) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Remove empty line

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java [hunk 6]
```diff
@@ -79,6 +79,9 @@ public void testGetByZoneAndSubjectIdentifier() {
                 .getByZoneAndSubjectIdentifier(TEST_ZONE_2, subjectEntityForZone2.getSubjectIdentifier());
         assertThat(actualSubjectForZone1, equalTo(subjectEntityForZone1));
         assertThat(actualSubjectForZone2, equalTo(subjectEntityForZone2));
+        this.subjectRepository.delete(subjectEntityForZone1);
+        this.subjectRepository.delete(subjectEntityForZone2);
+
```
Comment (anubhavi25) [inline service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Remove empty line

**Review by anubhavi25 — CHANGES_REQUESTED** [review]
@FrankGasparovic if you worked on this with @VoollaSandeepKumar , amend the commit with `git duet-commit` to add the sign off line

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java (hunk 4 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Fixed

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java (hunk 5 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Fixed

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java (hunk 3 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Fixed

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java (hunk 6 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphSubjectRepositoryTest.java]: Fixed

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java (hunk 2 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: Fixed

File: service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java (hunk 1 above)
Comment (FrankGasparovic) [reply service/src/test/java/com/ge/predix/acs/privilege/management/dao/GraphResourceRepositoryTest.java]: Fixed

**Comment by anubhavi25** [comment]
[Successful builds: #782-784,786](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/)

**Review by anubhavi25 — APPROVED** [review]

**Pull request merged**